### DISCUSSION TITLE
Multi line ternaries should be allowed in our style guide.

### DISCUSSION THREAD

**Discussion by nickkaczmarek** [wwt/SwiftCurrent#188] [General]
We talked a lot about this on slack so I'm going to copy that out and paste it here for clarity.

**tl;dr making multi line ternaries allowed but saying that they should be avoided if the use case doesn't really call for them ie don't get too complex.**

---
### An example:
![Screen Shot 2022-03-07 at 10 56 00 AM](https://user-images.githubusercontent.com/9044401/157084376-48d6c549-e6f5-4abc-ab32-609645eda308.png)

### Discussion

[Tyler Thompson](https://app.slack.com/team/U318QDK8A)  [20 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672207237609?thread_ts=1646672181.801119&cid=C021W3A5RFW)
Gonna respond here so we can get concrete.

[Tyler Thompson](https://app.slack.com/team/U318QDK8A)  [18 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672304436819?thread_ts=1646672181.801119&cid=C021W3A5RFW)
So here's the thing. type.type == objectType (when objectType is optional) is NOT semantically equivalent to.

```
if let objectType = objectType {
    return type.type == objectType
}
```
We've run into that problem before. The ternary relies on underlying implementation of type.type. However, what if both are nil? Did you mean to return true?

[Nick Kaczmarek](https://app.slack.com/team/U01D8V9NLLR):jupiter:  [18 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672327595899?thread_ts=1646672181.801119&cid=C021W3A5RFW)
I think this is an example of readibility, although I am not in love with this particular chunk of code. I do however find it easy enough to read. I think we could endeavor to extract the bools to make it even easier, but I think the one liner is fine for really short things but I think longer ones have a place too. Also I don't mind nested ternaries typically

[Tyler Thompson](https://app.slack.com/team/U318QDK8A)  [17 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672354270079?thread_ts=1646672181.801119&cid=C021W3A5RFW)
This logical statement is much easier to understand:
```
guard type.inhertiences.contains(conformance) else { return false }

if let objectType = objectType {
    return type.type == objectType
}
```
Intent is clear, side-effects are minimal.

[Tyler Thompson](https://app.slack.com/team/U318QDK8A)  [17 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672390406539?thread_ts=1646672181.801119&cid=C021W3A5RFW)
That entire thing can be returned as an expression by using a closure:
```
let conformanceCheck = {
    guard type.inhertiences.contains(conformance) else { return false }

    if let objectType = objectType {
        return type.type == objectType
    }
}()
```

[Nick Kaczmarek](https://app.slack.com/team/U01D8V9NLLR):jupiter:  [16 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672396788789?thread_ts=1646672181.801119&cid=C021W3A5RFW)
I think it's important (in this case) to not judge the code but judge the structure

[Tyler Thompson](https://app.slack.com/team/U318QDK8A)  [16 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672428295429?thread_ts=1646672181.801119&cid=C021W3A5RFW)
I am trying to judge the structure. My point was on intent. Readability is first and foremost about understanding intent of the author.

[Nick Kaczmarek](https://app.slack.com/team/U01D8V9NLLR):jupiter:  [16 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672430843769?thread_ts=1646672181.801119&cid=C021W3A5RFW)
It could be written in a more concise way, but using ternaries in this way is agreeable to me. I'll have to try to contrive an example, but that might take me a little while

[Tyler Thompson](https://app.slack.com/team/U318QDK8A)  [15 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672469120209?thread_ts=1646672181.801119&cid=C021W3A5RFW)
The ternary does not convey intent to me. Is my conformance check correct? Or did Morgan desire different behavior? I genuinely don't know the answer to that question.

[Nick Kaczmarek](https://app.slack.com/team/U01D8V9NLLR):jupiter:  [13 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672631496199?thread_ts=1646672181.801119&cid=C021W3A5RFW)
```
func something() {
    let ifThisIsTrue = Bool.random()
    let anotherThing = ifThisIsTrue ? "tanks" : "thanks"

    let anotherThing1 = ifThisIsTrue
        ? "tanks"
        : "thanks"

    let anotherThing2: String

    if ifThisIsTrue {
        anotherThing2 = "tanks"
    } else {
        anotherThing2 = "thanks"
    }

    print(anotherThing2, anotherThing, anotherThing1)
}
```

[Nick Kaczmarek](https://app.slack.com/team/U01D8V9NLLR):jupiter:  [12 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672667655849?thread_ts=1646672181.801119&cid=C021W3A5RFW)
Maybe I'm not thinking swiftly enough, but I find the first and middle options to be nice, but with a longer version of the second, I prefer the indentation.

[Nick Kaczmarek](https://app.slack.com/team/U01D8V9NLLR):jupiter:  [12 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672698546129?thread_ts=1646672181.801119&cid=C021W3A5RFW)
super contrived, but this is about the style for me. I think the intent is clear in my example.

[Morgan Zellers](https://app.slack.com/team/UAGTEPY06)  [12 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672718575689?thread_ts=1646672181.801119&cid=C021W3A5RFW)
So I’d prefer to stay away from analyzing the code around the ternary here. Maybe this is a bad example for the question we’re asking

[Tyler Thompson](https://app.slack.com/team/U318QDK8A)  [11 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672781812579?thread_ts=1646672181.801119&cid=C021W3A5RFW)
We can stay away, but I think it's a perfectly cogent example because it showed the point, multi-line ternaries opened the door to ambiguity. Your code is fine, I'm not saying you did it wrong. I'm saying that STYLE promoted code that left questions.

[Nick Kaczmarek](https://app.slack.com/team/U01D8V9NLLR):jupiter:  [11 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672783394809?thread_ts=1646672181.801119&cid=C021W3A5RFW)
That's why I try to post my super contrived example. So the discussion is less on the example and more on the usage of a ternary.

[Nick Kaczmarek](https://app.slack.com/team/U01D8V9NLLR):jupiter:  [10 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672810909279?thread_ts=1646672181.801119&cid=C021W3A5RFW)
I think I see where [@Tyler.Thompson](https://wwt.slack.com/team/U318QDK8A) is coming from, but I'm curious if my example changes your mind in any way

[Tyler Thompson](https://app.slack.com/team/U318QDK8A)  [10 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672829608959?thread_ts=1646672181.801119&cid=C021W3A5RFW)
With that example my argument is that the first example inherently encourages you to keep your ternary brief.
With your middle option while I have no objection to this code, but it opens the door to abuse. It's easier for me to decide to nest them, or add WAY too much in there.
Your final example is also fine, although I actually hate it the most because it's needlessly verbose.

[Nick Kaczmarek](https://app.slack.com/team/U01D8V9NLLR):jupiter:  [10 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672836846379?thread_ts=1646672181.801119&cid=C021W3A5RFW)
it's less that I'm saying to always use a ternary and more that I think there is value in a multiline version. Perhaps your point about AVOID is apt here

[Tyler Thompson](https://app.slack.com/team/U318QDK8A)  [9 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672891291099?thread_ts=1646672181.801119&cid=C021W3A5RFW)
Right, that's really what I'm suggesting here. It's AVOID. If given a choice, option 1 was the best. It's clear, it's concise, it encourages readability. If people start to multiline it's not inherently evil but it probably indicates there's something else going on.

[Nick Kaczmarek](https://app.slack.com/team/U01D8V9NLLR):jupiter:  [9 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672912840539?thread_ts=1646672181.801119&cid=C021W3A5RFW)
I think what I'm getting at is that there is a place for them. I agree that they can be abused, but I want to be able to write them if it makes sense. I think we still reserve the right to say that is a bad use case, make it better but that that won't always be the case.

[Nick Kaczmarek](https://app.slack.com/team/U01D8V9NLLR):jupiter:  [8 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672921573759?thread_ts=1646672181.801119&cid=C021W3A5RFW)
Cool beans

[Tyler Thompson](https://app.slack.com/team/U318QDK8A)  [8 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672932934999?thread_ts=1646672181.801119&cid=C021W3A5RFW)
It's the same reason we put line limits on files. Who the hell cares how big a file is? The compiler doesn't (well..the Swift compiler kind of does...) We do it because large files usually indicate a smell or a cognitive overload

[Nick Kaczmarek](https://app.slack.com/team/U01D8V9NLLR):jupiter:  [8 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672940342689?thread_ts=1646672181.801119&cid=C021W3A5RFW)
I kinda wish we would've done this in a discussion, but the feedback loop is so tight on here.

[Nick Kaczmarek](https://app.slack.com/team/U01D8V9NLLR):jupiter:  [8 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672950897099?thread_ts=1646672181.801119&cid=C021W3A5RFW)
Agree on file size.

[Tyler Thompson](https://app.slack.com/team/U318QDK8A)  [8 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672961368389?thread_ts=1646672181.801119&cid=C021W3A5RFW)
multi line ternaries are the same IMO, they're not bad. They're a flag that there might be dragons.
:plusone:
1

[Nick Kaczmarek](https://app.slack.com/team/U01D8V9NLLR):jupiter:  [8 minutes ago](https://wwt.slack.com/archives/C021W3A5RFW/p1646672967138349?thread_ts=1646672181.801119&cid=C021W3A5RFW)
I might try to screen shot this and post it.

**Comment by nickkaczmarek** [comment]
I think we landed on 

**making multi line ternaries allowed but saying that they should be avoided if the use case doesn't really call for them ie don't get too complex.**
